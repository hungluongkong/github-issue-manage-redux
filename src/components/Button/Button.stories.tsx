import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Button from '.';
import { PreviewArea, Code, DetailTable } from '../commonStyled';
import { ButtonType, ButtonSize } from '../../types';

// Detail of props
const propsDetail: JSX.Element = DetailTable({
  data: [
    {
      name: 'type',
      type: 'enum ButtonType { Default, Primary }',
      default: 'ButtonType.Primary',
    },
    {
      name: 'size',
      type: 'enum ButtonSize { Small, Normal }',
      default: 'ButtonSize.Normal',
    },
    { name: 'value', type: 'string', default: 'ButtonType.Primary' },
    {
      name: 'onClicked',
      type: 'function',
      description:
        '((event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void)',
    },
  ],
});

storiesOf('Button', module)
  .addDecorator((story) => <div style={{ margin: '0 30px' }}>{story()}</div>)
  .add('Button Style', () => (
    <>
      <h3>Button Types</h3>
      <PreviewArea>
        <Button
          type={ButtonType.Primary}
          onClick={action('click')}
          value="Primary Button"
        />
        <Button
          type={ButtonType.Default}
          onClick={action('click')}
          value="Default Button"
        />
      </PreviewArea>
      <Code>
        {`
          <Button type={ButtonType.Primary} onClick={action('click')}>Primary Button</Button>
          <Button type={ButtonType.Default} onClick={action('click')}>Default Button</Button>
        `}
      </Code>
      {propsDetail}
    </>
  ))
  .add('Button Size', () => (
    <>
      <h3>Button Size</h3>
      <PreviewArea>
        <Button
          onClick={action('click')}
          size={ButtonSize.Normal}
          value="Normal Button"
        />
        <Button
          onClick={action('click')}
          size={ButtonSize.Small}
          value="Small Button"
        />
      </PreviewArea>
      <Code>
        {`
          <Button onClick={action('click')} size={ButtonSize.Normal}>Normal Button</Button>
          <Button onClick={action('click')} size={ButtonSize.Small}>Small Button</Button>
        `}
      </Code>
      {propsDetail}
    </>
  ));
