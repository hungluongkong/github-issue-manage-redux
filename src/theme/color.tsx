export const PRIMARY_COLOR = 'rgb(26, 170, 85)';
export const DEFAULT_COLOR = 'rgb(255, 255, 255)';

export const INPUT_BG_COLOR = 'rgb(253, 253, 253)';
export const INPUT_BORDER_COLOR = 'rgb(225, 228, 232)';
export const SHADOW_BORDER = 'rgba(145, 92, 182, 0.3)';

export const CODE_BG_COLOR = 'rgb(241, 241, 241)';

export const SMALL_TEXT_COLOR = 'rbg(88, 96, 105)';

export const ISSUE_BG_DEFAULT = 'rgb(205, 247, 227)';
export const ISSUE_BG_LOCKED = 'rgb(204, 217, 211)';
export const ISSUE_BG_CONTENT = 'rgb(169, 195, 204)';

export const ISSUE_LIST_BG = 'rgb(235, 245, 255)';
export const ISSUE_LIST_HEAD_BG = 'rgb(240, 240, 240)';

export const LOADING_BAR_COLOR = 'rgb(219, 186, 238)';

export const POPUP_BORDER_COLOR = 'rgba(0, 0, 0, 0.2)';
export const POPUP_BG_COLOR_DEFAULT = DEFAULT_COLOR;
export const POPUP_BG_COLOR_SUCCESS = PRIMARY_COLOR;
export const POPUP_BG_COLOR_WARNING = 'rgb(235, 216, 124)';
