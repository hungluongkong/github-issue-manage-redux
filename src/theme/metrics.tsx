/* ------------------ Metrics ----------------- */

export const OPACITY = {
  default: 1,
  hover: 0.7,
};

export const ButtonSizing = {
  default: {
    width: '101px',
    lineHeight: '20px',
  },
  small: {
    width: '101px',
    lineHeight: '15px',
  },
};

export const ZINDEX = {
  Default: 1,
  Popup: 10,
};

export const POPUP_TIMEOUT = 2000;
/* ------------------- Font ------------------- */

export const FontSize = {
  Default: '0.875rem',
  Small: '0.75rem',
  Icon: '1.25rem',
  Code: '0.75rem',
  Title: '1.875rem',
};

export const FontFamily = {
  Default: `-apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
            'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
            sans-serif`,
  Code: 'Consolas,"courier new"',
};
