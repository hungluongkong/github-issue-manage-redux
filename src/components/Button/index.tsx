import React from 'react';
import styled from 'styled-components';

import * as color from '../../theme/color';
import * as metrics from '../../theme/metrics';
import { ButtonType, ButtonProps, ButtonSize } from '../../types';

/**
 * Get css style for each type of button.
 * @param type type of button (enum)
 */
const typeToColor = (type: ButtonType): string =>
  type === ButtonType.Primary
    ? `
      background: ${color.PRIMARY_COLOR};
      color: ${color.DEFAULT_COLOR};
    `
    : `
      background: ${color.DEFAULT_COLOR};
      color: ${color.PRIMARY_COLOR};
    `;

/**
 * Styling button by buttonSize
 * @param bSize size of button (enum)
 */
const sizeToSize = (bSize: ButtonSize): string =>
  bSize === ButtonSize.Small
    ? `
      line-height: ${metrics.ButtonSizing.small.lineHeight};
      padding: 3px 5px;
      font-size: ${metrics.FontSize.Small};
    `
    : // Style normal button here
      '';

const StyledButton = styled.button<{ bType: ButtonType; size: ButtonSize }>`
  display: inline-block;
  min-width: 101px;
  text-align: center;
  font-size: ${metrics.FontSize.Default};
  line-height: 20px;
  padding: 6px 10px;
  vertical-align: middle;
  border-radius: 4px;
  border: ${color.PRIMARY_COLOR} 1px solid;

  :hover {
    opacity: ${metrics.OPACITY.hover};
    box-shadow: 0 5px 15px ${color.SHADOW_BORDER};
    cursor: pointer;
  }

  ${(props: { bType: ButtonType }): string => typeToColor(props.bType)}
  ${(props: { size: ButtonSize }): string => sizeToSize(props.size)}
`;

const Button = ({
  type = ButtonType.Primary,
  size = ButtonSize.Normal,
  value = '',
  onClick = () => {},
}: ButtonProps): JSX.Element => (
  <StyledButton
    {...{ size, onClick }}
    // bType: button type
    bType={type}
  >
    {value}
  </StyledButton>
);

export default Button;
