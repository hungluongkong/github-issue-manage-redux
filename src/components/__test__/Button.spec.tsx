import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { shallow } from 'enzyme';
import Button from '../Button';

import * as color from './../../theme/color';

describe('Unit test for Button component', () => {
  it('snapshot', () => {
    const buttonComponent = renderer.create(<Button value="Button" />);
    expect(buttonComponent).toMatchSnapshot();
  });

  it('simulate click event', () => {
    const mockClick = jest.fn();
    const button = shallow(<Button onClick={mockClick} value="Button" />);

    button.simulate('click');
    expect(mockClick).toHaveBeenCalledTimes(1);
  });
});
