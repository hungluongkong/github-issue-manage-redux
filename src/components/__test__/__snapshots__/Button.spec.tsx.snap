// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for Button component snapshot 1`] = `
.c0 {
  display: inline-block;
  min-width: 101px;
  text-align: center;
  font-size: 0.875rem;
  line-height: 20px;
  padding: 6px 10px;
  vertical-align: middle;
  border-radius: 4px;
  border: rgb(26,170,85) 1px solid;
  background: rgb(26,170,85);
  color: rgb(255,255,255);
}

.c0:hover {
  opacity: 0.7;
  box-shadow: 0 5px 15px rgba(145,92,182,0.3);
  cursor: pointer;
}

<button
  className="c0"
  onClick={[Function]}
  size={1}
>
  Button
</button>
`;
